### PR TITLE
[distributed sampling] introduce new priority sampler for distributed sampling

### PR DESCRIPTION
**Pull request by ufoot**
Follow-up to #325 which isolates the generic, core features, from contrib specific enhancements which may make sense without priority sampling.

### CODE REVIEW

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 1]
```diff
@@ -29,24 +31,30 @@ def attach_context(request):
         service = app[CONFIG_KEY]['service']
         distributed_tracing = app[CONFIG_KEY]['distributed_tracing_enabled']
 
+        context = tracer.get_call_context()
+
+        # Create a new context based on the propagated information
+        # Do not fill context with distributed sampling if the tracer is disabled
+        # because the would call the callee to generate references to data which
+        # has never been sent to agent.
+        if tracer.enabled and distributed_tracing:
```
Comment (LotharSee) [inline ddtrace/contrib/aiohttp/middlewares.py]: I'd suggest to add a TODO here.

We could clearly have a helper somewhere which does http_headers -> new_context. As this will be reusable for all HTTP-related libs.

File: ddtrace/contrib/asyncio/provider.py [hunk 2]
```diff
@@ -14,8 +14,28 @@ class AsyncioContextProvider(DefaultContextProvider):
     execution. It must be used in asynchronous programming that relies
     in the built-in ``asyncio`` library. Framework instrumentation that
     is built on top of the ``asyncio`` library, can use this provider.
+
+    This Context Provider inherits from ``DefaultContextProvider`` because
+    it uses a thread-local storage when the ``Context`` is propagated to
+    a different thread, than the one that is running the async loop.
     """
-    def __call__(self, loop=None):
+    def activate(self, context, loop=None):
```
Comment (LotharSee) [inline ddtrace/contrib/asyncio/provider.py]: Is it normal that this context API change (adding the `activate `) is part of this PR?
Could we merge the context API change to simplify this PR a bit?

cc @palazzem

File: ddtrace/tracer.py [hunk 3]
```diff
@@ -80,13 +90,16 @@ def configure(self, enabled=None, hostname=None, port=None, sampler=None,
...
-        :param object sampler: A custom Sampler instance
+        :param object sampler: A custom Sampler instance, locally deciding to totally drop the trace or not.
+        :param object priority_sampler: A custom Sampler instance, taking the priority sampling decision.
         :param object context_provider: The ``ContextProvider`` that will be used to retrieve
             automatically the current call context. This is an advanced option that usually
             doesn't need to be changed from the default value
         :param object wrap_executor: callable that is used when a function is decorated with
             ``Tracer.wrap()``. This is an advanced option that usually doesn't need to be changed
             from the default value
+        :param priority_sampling: enable priority sampling, this is required for
```
Comment (LotharSee) [inline ddtrace/tracer.py]: It makes the API a bit more complicated to have both `priority_sampler` and `priority_sampling`.

I guess the purpose was to simply allow to set `priority_sampling ` to turn it on?
Then I'd suggest to only keep this one. As no one will have to alter the `priority_sampler` ; let's only have `priority_sampling` as part of the API.

File: ddtrace/contrib/asyncio/provider.py (hunk 2 above)
Comment (palazzem) [reply ddtrace/contrib/asyncio/provider.py]: merged the other one so this should not be a problem anymore

File: ddtrace/span.py [hunk 4]
```diff
@@ -181,6 +184,31 @@ def set_metrics(self, metrics):
...
+        if sampling_priority is None:
+            self.priority = None
+        else:
+            try:
+                self.priority = int(sampling_priority)
+            except ValueError:
+                # if the provided sampling_priority is invalid, ignore it.
+                pass
+
+    def get_sampling_priority(self):
```
Comment (palazzem) [inline ddtrace/span.py]: Probably we need to change that. The pythonic way is to have a property instead of getter and setter (otherwise we'll finish having getter and setter for everything). Low priority, we can refactor that before the merge.

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 5]
```diff
@@ -29,24 +31,30 @@ def attach_context(request):
...
+            sampling_priority = request.headers.get(SAMPLING_PRIORITY_HEADER_ID)
+            # keep sampling priority as None if not propagated, to support older client versions on the parent side
+            if sampling_priority:
+                sampling_priority = int(sampling_priority)
+
+            context = Context(trace_id=trace_id, span_id=parent_span_id, sampling_priority=sampling_priority)
+
         # trace the handler
-        request_span = tracer.trace(
+        request_span = tracer.start_span(
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/aiohttp/middlewares.py]: Will write that test, indeed, required.

File: ddtrace/tracer.py [hunk 6]
```diff
@@ -174,24 +202,34 @@ def start_span(self, name, child_of=None, service=None, resource=None, span_type
...
+                    # If dropped by the local sampler, distributed instrumentation can drop it too.
+                    span.set_sampling_priority(0)
 
         # add common tags
         if self.tags:
             span.set_tags(self.tags)
+        if not span._parent:
+            span.set_tag(system.PID, getpid())
+
+        # TODO: add protection if the service is missing?
```
Comment (palazzem) [inline ddtrace/tracer.py]: yes but it's another topic that would be handled by our compatibility layer. You can safely remove this `TODO` since I already track this.

File: ddtrace/context.py [hunk 7]
```diff
@@ -50,13 +56,28 @@ def get_current_span(self):
...
+        Set current span internally.
+
+        Non-safe if not used with a lock. For internal Context usage only.
+        """
+        self._current_span = span
+        if span:
+            self._parent_trace_id = span.trace_id
+            self._parent_span_id = span.span_id
+            self._sampled = span.sampled
+            self._sampling_priority = span.get_sampling_priority()
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/context.py]: Will use the pythonic approach of having a property, AFAIR the getter/setter were here initially because first implementations were actually setting a meta so using this did save a lot of copy/paste. But now that we have this sampling_priority as a class member, they sound obviously over engineered. Removing them.

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 8]
```diff
@@ -29,24 +31,32 @@ def attach_context(request):
         service = app[CONFIG_KEY]['service']
         distributed_tracing = app[CONFIG_KEY]['distributed_tracing_enabled']
 
+        context = tracer.context_provider.active()
+
+        # Create a new context based on the propagated information.
+        # Do not fill context with distributed sampling if the tracer is disabled
+        # because the would call the callee to generate references to data which
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/aiohttp/middlewares.py]: OK this is quite fundamental, let me rephrase it then.

[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/aiohttp/middlewares.py]: OK, this really does not apply here so I'm removing the test on `tracer.enabled`. But this would be very important to test for a client. On the server side we don't care as we're free to report whatever we want. But if acting as a client (http client for instance) we should not set the trace ID, parent ID, or distributed sampling attributes if the tracer is disabled. Because then, the server we're calling would think he is responding to a parent trace, but this trace will never be sent to the agent, so it can only result in broken traces.

File: ddtrace/context.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/context.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/e1257b411df64ab032a30b00315d55a373d1f4a3

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 9]
```diff
@@ -11,6 +12,7 @@
 
 PARENT_TRACE_HEADER_ID = 'x-datadog-trace-id'
 PARENT_SPAN_HEADER_ID = 'x-datadog-parent-id'
+SAMPLING_PRIORITY_HEADER_ID = 'x-datadog-sampling-priority'
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/aiohttp/middlewares.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/9924dbf293dfbdd1d6cda2e28e44a90b1f20faa9

File: ddtrace/contrib/aiohttp/middlewares.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/aiohttp/middlewares.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/858a4f014218df42975dac22356f5a000fdb355c

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 10]
```diff
@@ -29,24 +31,32 @@ def attach_context(request):
...
+        # Create a new context based on the propagated information.
+        # Do not fill context with distributed sampling if the tracer is disabled
+        # because the would call the callee to generate references to data which
+        # has never been sent to agent.
+        # [TODO:christian] this is quite generic and applies to any similar library so
+        # at some point we should have some shared code which populates context from headers.
+        if tracer.enabled and distributed_tracing:
+            trace_id = int(request.headers.get(PARENT_TRACE_HEADER_ID, 0))
+            parent_span_id = int(request.headers.get(PARENT_SPAN_HEADER_ID, 0))
+            sampling_priority = request.headers.get(SAMPLING_PRIORITY_HEADER_ID)
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/aiohttp/middlewares.py]: Sure? Because what we want is:

- header set to 1 -> sampling_priority == 1
- header set to 0 -> sampling_priority == 0
- header unset -> sampling_priority == None

But with this one liner I think if the header is set to 0 we get request.headers.get(...,0) -> this is 0, so then the or evaluates its right member and returns None (instead of 0). I tried it, it breaks a unitary test. Having sampling_priority to None or 0 is a very different case, when None, it means we should use the previous signature sampler, if set to 0 it means the trace should not be sent according to priority sampling policy.

Thanks however for zooming on that point, it happens there's a little caveat below. Updating.

File: ddtrace/contrib/aiohttp/middlewares.py [hunk 11]
```diff
@@ -29,24 +31,32 @@ def attach_context(request):
...
+            # keep sampling priority as None if not propagated, to support older client versions on the parent side
+            if sampling_priority:
```
Comment (ufoot) [inline ddtrace/contrib/aiohttp/middlewares.py]: This test is wrong it should be `is not None`

Comment (ufoot) [reply ddtrace/contrib/aiohttp/middlewares.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/92ed47f4848194509f6682b7240b19c8883d4b81

File: ddtrace/contrib/asyncio/helpers.py [hunk 12]
```diff
@@ -74,7 +77,7 @@ def _wrap_executor(fn, args, tracer, ctx):
     # the AsyncioContextProvider knows that this is a new thread
     # so it is legit to pass the Context in the thread-local storage;
     # fn() will be executed outside the asyncio loop as a synchronous code
-    tracer._context_provider._local.set(ctx)
+    tracer.context_provider.activate(ctx)
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/asyncio/helpers.py]: OK, in practice both work so it's the about the one you prefer maybe @palazzem , switching back to `tracer.context_provider.activate(ctx)` https://github.com/DataDog/dd-trace-py/pull/359/commits/683e8fd64828f9259b233ebab57e279e56b9f89b

File: ddtrace/contrib/asyncio/helpers.py [hunk 13]
```diff
@@ -77,7 +77,7 @@ def _wrap_executor(fn, args, tracer, ctx):
     # the AsyncioContextProvider knows that this is a new thread
     # so it is legit to pass the Context in the thread-local storage;
     # fn() will be executed outside the asyncio loop as a synchronous code
-    tracer.context_provider.activate(ctx)
+    tracer._context_provider._local.set(ctx)
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/asyncio/helpers.py]: Switching back to `tracer.context_provider.activate(ctx)` https://github.com/DataDog/dd-trace-py/pull/359/commits/683e8fd64828f9259b233ebab57e279e56b9f89b

[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/contrib/asyncio/helpers.py]: ^ I'm not able to find a edge case that would have this fail BTW :-/

File: ddtrace/contrib/aiohttp/middlewares.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (palazzem) [reply ddtrace/contrib/aiohttp/middlewares.py]: oh yes that makes sense! will keep track of this for another PR in `requests` and `httplib`

File: ddtrace/contrib/aiohttp/middlewares.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (palazzem) [reply ddtrace/contrib/aiohttp/middlewares.py]: Oh in that case yes, the oneliner doesn't apply. Keep it that way so.

File: ddtrace/sampler.py [hunk 14]
```diff
@@ -44,59 +46,67 @@ def set_sample_rate(self, sample_rate):
...
     """
 
-    # Reasonable values
-    BUCKETS_PER_S = 10
-    BUFFER_DURATION = 2
-    BUFFER_SIZE = BUCKETS_PER_S * BUFFER_DURATION
+    def __init__(self, sample_rate=1):
+        self._lock = Lock()
+        self._by_service_samplers = {}
+        self._default_key = self._key(None, None)
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/sampler.py]: OK, will refactor with `()` indeed :+1:. The use cases for using the default keys can be:

- a broken (custom or in-house) instrumentation which is now setting the service early enough. In that case it has no way to guess what rate to apply, because when they are submitted later, the spans do have a service set. So the idea here is to pick the default value returned by the agent, which is the global sampling rate. Technically, it causes over-sampling (dropping too many traces), which is not great, but not as bad as using `1` and send everything.
- an application that is starting and/or is having some service name change (services could be defined dynamically according to some variable flag, who knows, or some services could show up rarely and appear "some time" after the other) so when a new service shows up, instead of using 1 we use the default sampling rate (this is done here: https://github.com/DataDog/dd-trace-py/pull/359/files#diff-073d17b74132da4920e45dbb8d5de860R88) which at least is a little context dependant, if we're already at `0.01` we won't "jump" to `1` just because the service is new.

Globally, it's really about providing a default behavior that is a little context dependant, it's not the right rate (this, we only now later) but at least we get a reasonable hint on what to do.

File: ddtrace/sampler.py [hunk 15]
```diff
@@ -44,59 +46,67 @@ def set_sample_rate(self, sample_rate):
...
+        with self._lock:
+            if key in self._by_service_samplers:
+                return self._by_service_samplers[key].sample(span)
+            return self._by_service_samplers[self._default_key].sample(span)
+
+    def set_sample_rates_from_json(self, body):
+        log.debug("setting sample rates from JSON '%s'" % repr(body))
+        try:
+            if not isinstance(body, str):
+                body = body.decode('utf-8')
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/sampler.py]: :+1: refactoring

File: ddtrace/sampler.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/sampler.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/10eac176d05725d5450c2c5bfa8a1acd7f002224

File: ddtrace/sampler.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/sampler.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/cc47f23132845fc68c1038e3c3ed4eda0ed6dfc9

File: ddtrace/writer.py [hunk 16]
```diff
@@ -151,6 +156,11 @@ def _target(self):
                 # no traces and the queue is closed. our work is done
                 return
 
+            if hasattr(result_traces, 'read'):
+                result_traces_body = result_traces.read()
+                if hasattr(self._priority_sampler, 'set_sample_rates_from_json'):
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/writer.py]: https://github.com/DataDog/dd-trace-py/commit/cc47f23132845fc68c1038e3c3ed4eda0ed6dfc9 makes this obsolete. AFAIK it was here initially because I had plans to technically support users overriding the sampler with there own impl. As now we only support `priority_sampling=True` and not `priority_sampler=MyOwnSampler` we're fine.

File: tests/test_sampler.py [hunk 17]
```diff
@@ -34,121 +33,93 @@ def test_sample_rate_deviation(self):
...
+        for sample_rate in [0.1, 0.25, 0.5, 1]:
+            tracer = Tracer()
+            tracer.configure(sampler=AllSampler(), priority_sampling=True)
+            tracer.priority_sampler.set_sample_rate(sample_rate)
+            tracer.writer = writer
 
-        # Let's sample to a multiple of BUFFER_SIZE, so that we can pre-populate the buffer
-        tps = 15 * ThroughputSampler.BUFFER_SIZE
-        tracer.sampler = ThroughputSampler(tps)
+            random.seed(1234)
```
[the remark this answers is not in the document]
Comment (ufoot) [reply tests/test_sampler.py]: To reduce flakyness, indeed. A througput sampler is based on trace IDs which are pseudo-random so indeed with bad luck even if tuned at say 3% it could behave like it's 30%. This should be rare but again, that's the definition of flakiness.

File: tests/test_sampler.py [hunk 18]
```diff
@@ -34,121 +33,93 @@ def test_sample_rate_deviation(self):
...
+        }
 
-        got = len(traces)
-        expected = tps * total_time
-        error_delta = tps * ThroughputSampler.BUFFER_DURATION
+        writer = DummyWriter()
 
-        assert abs(got - expected) <= error_delta, \
-            "Wrong number of traces sampled, %s instead of %s (error_delta > %s)" % (got, expected, error_delta)
+        tracer = Tracer()
```
[the remark this answers is not in the document]
Comment (ufoot) [reply tests/test_sampler.py]: https://github.com/DataDog/dd-trace-py/commit/9b4d86fdd3729b504f86f402b7ad38bc123546aa should fix it (contains other refactorings related to your comments above)

File: ddtrace/sampler.py [hunk 19]
```diff
@@ -44,59 +46,70 @@ def set_sample_rate(self, sample_rate):
...
+        tags = span.tracer().tags
+        env = tags['env'] if 'env' in tags else None
+        key = _key(span.service, env)
+        with self._lock:
+            if key in self._by_service_samplers:
+                return self._by_service_samplers[key].sample(span)
+            return self._by_service_samplers[_default_key].sample(span)
+
+    def set_sample_rate_by_service(self, rate_by_service):
+        # log.debug("setting sample rates with '%s'" % repr(rates))
```
Comment (palazzem) [inline ddtrace/sampler.py]: I think we should remove these comments

File: ddtrace/span.py [hunk 20]
```diff
@@ -214,6 +217,12 @@ def to_dict(self):
         if self.span_type:
             d['type'] = self.span_type
 
+        if self._sampling_priority is not None:
```
Comment (palazzem) [inline ddtrace/span.py]: if that attribute stays "internal" it's fine to not use functions / properties; do we plan to never expose that to users?

File: tests/contrib/aiohttp/app/web.py [hunk 21]
```diff
@@ -54,6 +54,15 @@ def nested():
 
 
 @asyncio.coroutine
+def route_sub_span(request):
+    tracer = get_tracer(request)
```
[the remark this answers is not in the document]
Comment (palazzem) [reply tests/contrib/aiohttp/app/web.py]: @ufoot should we call `get_tracer` twice?

File: tests/test_integration.py [hunk 22]
```diff
@@ -480,3 +480,41 @@ def test_downgrade_api(self):
...
+        self.api_json = API('localhost', 8126, encoder=JSONEncoder())
+        self.api_msgpack = API('localhost', 8126, encoder=MsgpackEncoder())
+
+    def test_send_single_trace(self):
+        # register a single trace with a span and send them to the trace agent
+        self.tracer.trace('client.testing').finish()
+        trace = self.tracer.writer.pop()
+        traces = [trace]
+
+        # [TODO:christian] when CI has an agent that is able to process the v0.4
```
Comment (palazzem) [inline tests/test_integration.py]: ok will keep track of that

File: tests/test_sampler.py [hunk 23]
```diff
@@ -2,30 +2,26 @@
...
         for sample_rate in [0.1, 0.25, 0.5, 1]:
-            tracer = Tracer()
-            tracer.writer = writer
+            tracer = get_dummy_tracer()
+            writer = tracer.writer
 
-            sample_rate = 0.5
             tracer.sampler = RateSampler(sample_rate)
 
             random.seed(1234)
```
Comment (palazzem) [inline tests/test_sampler.py]: oh it was already here

File: ddtrace/sampler.py (hunk 19 above)
Comment (ufoot) [reply ddtrace/sampler.py]: :+1:

File: ddtrace/span.py (hunk 20 above)
Comment (ufoot) [reply ddtrace/span.py]: My position is: right now we use a `_sampling_priority_v1` metric instead of the `sampling.priority` tag that would ensure OT compatibility, so my guess is all this is beta, therefore fine to keep private, we can make it explicitly public later. There are subtles implications of using an attribute it, one is: it's fine to set the service/error state later than span creation as long as no distributed child span has been created, this implies control on when `_sampling_priority` is accessed. Similarly, setting `_sampling_priority_v1` after a child, distributed trace has been triggered, will lead to clear plain bugs as the information has already be sent. All in all, this is really a use-at-your-risk member, we might some day harden it, but my expectation this should come along with OT support, typically.

File: ddtrace/api.py [hunk 24]
```diff
@@ -12,40 +13,89 @@
...
+                body = body.decode('utf-8')
+            if body.startswith('OK'):
+                # This typically happens when using a priority-sampling enabled
+                # library with an outdated agent. It still works, but priority sampling
+                # will probably send too many traces, so the next step is to upgrade agent.
+                log.debug("'OK' is not a valid JSON, please make sure trace-agent is up to date")
+                return
+            content = loads(body)
+            return content
+        except ValueError as err:
```
[the remark this answers is not in the document]
Comment (ufoot) [reply ddtrace/api.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/bf4e369828989a89a6c8b444cc894bca93712f56

File: tests/test_integration.py (hunk 22 above)
Comment (ufoot) [reply tests/test_integration.py]: https://github.com/DataDog/dd-trace-py/pull/359/commits/1812ba0c8966f227a47afe0dd73ef217fa19b42c

File: ddtrace/span.py (hunk 20 above)
Comment (palazzem) [reply ddtrace/span.py]: totally agree! also this simplifies the implementation

File: ddtrace/tracer.py [hunk 25]
```diff
@@ -76,7 +79,8 @@ def context_provider(self):
         return self._context_provider
 
     def configure(self, enabled=None, hostname=None, port=None, sampler=None,
-                context_provider=None, wrap_executor=None, settings=None):
+                  context_provider=None, wrap_executor=None, priority_sampling=None,
```
Comment (ufoot) [inline ddtrace/tracer.py]: By default, priority sampling is disabled, it needs to be turned on explicitly by setting `priority_sampling` to `True`.

File: ddtrace/ext/distributed.py [hunk 26]
```diff
@@ -0,0 +1,6 @@
+# HTTP headers one should set for distributed tracing.
+# These are cross-language (eg: Python, Go and other implementations should honor these)
+
+HTTP_HEADER_TRACE_ID = 'x-datadog-trace-id'
+HTTP_HEADER_PARENT_ID = 'x-datadog-parent-id'
+HTTP_HEADER_SAMPLING_PRIORITY = 'x-datadog-sampling-priority'
```
Comment (ufoot) [inline ddtrace/ext/distributed.py]: This should be the same across all HTTP implementations of distributed tracing, for any language (Ruby, Java, Go, C#...)

File: ddtrace/tracer.py [hunk 27]
```diff
@@ -174,24 +202,35 @@ def start_span(self, name, child_of=None, service=None, resource=None, span_type
...
-            #    http://pypi.datadoghq.com/trace/docs/#distributed-tracing
-            parent_trace_id, parent_span_id = context._get_parent_span_ids()
-            if parent_trace_id:
-                span.trace_id = parent_trace_id
-
-            if parent_span_id:
-                span.parent_id = parent_span_id
-
-            self.sampler.sample(span)
+            span.sampled = self.sampler.sample(span)
```
Comment (p-lambert) [inline ddtrace/tracer.py]: What is the point of having a common interface for all samplers if we care about the concrete implementation and do type checks outside of it?

I thought the original API made more sense before. If I'm a `RateSampler` I know how to annotate the span accordingly. Also, we could have instead a `PrioritySampler` that first delegates the sampling to to the `RateSampler` and if the span is sampled, it also calls `RateByServiceSampler` and annotates it.

That way, the `Tracer` calls only `sampler.sample(span)` and we don't break encapsulation.

`PrioritySampler` would looks like:
```py
class PrioritySampler(object):
    def sample(self, span):
        sampled = rate_sampler.sample(span)
        by_service_sample.sample(span) if sampled
        return sampled
```

Comment (ufoot) [reply ddtrace/tracer.py]: You're quite right. But note that `PrioritySampler` would not necessarily call a `RateSampler` as the signature default sampler can be overridden in the constructor and for that part, we *do not* do type checking (internally, we actually replace it by something else on occasions) so the priority sampler should still run "any sampler that has a simple sample() func" if it decided to do everything itself. Mostly, the more we make this private, the better, because probably we don't want much external code to rely on the current implementation.

Comment (p-lambert) [reply ddtrace/tracer.py]: @ufoot 
Oh, I see! If I understood correctly, we can still have something like the code below in the `.configure` method, right?
(so we let the underlying sampler be injected in the constructor)

```py
sampler = PrioritySampler(sampler) if priority_sampling
```

File: ddtrace/api.py [hunk 28]
```diff
@@ -12,40 +13,89 @@
...
+            return content
+        except (ValueError, TypeError) as err:
+            log.debug("unable to load JSON '%s': %s" % (body, err))
+
 class API(object):
     """
     Send data to the trace agent using the HTTP protocol and JSON format
     """
-    def __init__(self, hostname, port, headers=None, encoder=None):
+    def __init__(self, hostname, port, headers=None, encoder=None, priority_sampling=False):
```
Comment (p-lambert) [inline ddtrace/api.py]: I think if we let the caller of this method provide an API version instead of the `priority_sampling` flag, we would have a bit more flexibility (and we could also get rid of the conditional below).

Comment (palazzem) [reply ddtrace/api.py]: I think it's more a matter to simplify the caller logic, plus decoupling both components.

With this implementation, you [propagate](https://github.com/DataDog/dd-trace-py/pull/359/files#diff-c3dd6d04070c4b00a58e7ab4aeafd404R34) the `priority_sampling` boolean as a feature flag (available in the `.configure()` method), rather than requiring the `Writer` class to know which API version is the Priority Sampling API (that is known only by the `API` class).

Comment (palazzem) [reply ddtrace/api.py]: One more thing: all of this is internal machinery so not a public API people will play with.

File: ddtrace/context.py [hunk 29]
```diff
@@ -20,25 +20,31 @@ class Context(object):
...
         self._current_span = None
         self._lock = threading.Lock()
-        self._parent_span_id = span_id
+
         self._parent_trace_id = trace_id
+        self._parent_span_id = span_id
+        self._sampled = sampled
+        self._sampling_priority = sampling_priority
+
+    def get_context_attributes(self):
```
[the remark this answers is not in the document]
Comment (palazzem) [reply ddtrace/context.py]: @LotharSee as we've discussed, we're hiding this API with the Injector/Extractor API so that users generates a `Context` using something more similar to OpenTracing.
